CMakeLists.txt

```diff
@@ -0,0 +1,83 @@
+#
+# Copyright 2017 The Abseil Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+cmake_minimum_required(VERSION 2.8.12)
+project(absl)
+
+# enable ctest
+include(CTest)
+
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/CMake)
+
+include(GNUInstallDirs)
+include(AbseilHelpers)
+
+
+# config options
+set(ABSL_STD_CXX_FLAG "${CMAKE_CXX_STD_CXX11}" CACHE STRING "c++ std flag (default: c++11)")
+
+
+# include current path
+list(APPEND ABSL_COMMON_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR})
+
+# -std=X
+set(CMAKE_CXX_FLAGS "${ABSL_STD_CXX_FLAG} ${CMAKE_CXX_FLAGS}") 
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_WARNING_VLA} ${CMAKE_CXX_FLAGS} ")
+
+
+# find dependencies
+## pthread
+find_package(Threads REQUIRED)
+
+## check targets
+check_target(cctz)
+check_target(gtest)
+check_target(gtest_main)
+check_target(gmock)
+
+# -fexceptions
+set(ABSL_EXCEPTIONS_FLAG "${CMAKE_CXX_EXCEPTIONS}")
+
+# fix stuff
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FIX_MSVC} ${CMAKE_CXX_FLAGS}") 
+
+list(APPEND ABSL_TEST_COMMON_LIBRARIES
+  gtest_main
+  gtest
+  gmock
+  ${CMAKE_THREAD_LIBS_INIT}
+)
+
+add_subdirectory(absl)
+
+
+#
+#
+## public exported target absl::
+## please use these targets in your project
+#
+#
+add_library(absl::algorithm ALIAS absl_algorithm)
```

Comment:
This would be better done in the macros defining the targets.

Reply:
I did that on purpose to make explicit to the user what are the public target to use. I would prefer maintain it like that excepted if you have strong objections.

Reply:
At some point, someone will add a new target and forget to add the line in there. It's better if it's done by the macro automatically.
This type of information needs to go in documentation actually.

Reply:
I understand your point.
However, I am convince that 70% of your user will not even take the time to check the documentation like always :)

I fixed that. The alias is automatically generated in the macro if tagged explicitely, to allow private / public namespace and a comment is added in the main CMake file for the user.
I will also add this information in the doc